Add 'dbt Latest' dev blog (#5429)
fishtown-analytics/docs.getdbt.com: website/blog/2024-05-22-latest-dbt-stability-improvement-innovation.md

---
title: "Latest dbt: a stable foundation for improvement & innovation"
description: "Over the past 6 months, we've laid a stable foundation for continuously improving dbt."
slug: latest-dbt-stability

> authors: [michelle_ark, chenyu_li, colin_rogers]

Reviewer: @jtcohen6 shouldn't you be an author here as well? 😄

Author: i'm just the proud publisher ;)

(azzam too)

tags: [dbt Cloud]
hide_table_of_contents: false

date: 2024-05-02
is_featured: true
---

> As long as it's existed, dbt Cloud has required users to select a version of dbt Core to use under the hood in their jobs and environments. This made sense in the earliest days, when dbt Core minor versions often included breaking changes. It provided a clear way for everyone to know which version of the underlying runtime they were getting.

Reviewer 2: just sugg 'dbt cloud' at the beginning
```suggestion
As long as dbt Cloud existed, it has required users to select a version of dbt Core to use under the hood in their jobs and environments. This made sense in the earliest days when dbt Core minor versions often included breaking changes. It provided a clear way for everyone to know which version of the underlying runtime they were getting.
```

However, this came at a cost. While bumping a project's dbt version *appeared* as simple as selecting from a dropdown, there was real effort required to test the compatibility of the new version against existing projects, package dependencies, and adapters. On the other hand, putting this off meant foregoing access to new features and bug fixes in dbt.

> But no more. Today, we're ready to announce the general availability of a new option in dbt Cloud: **"Keep on latest version."** 

Reviewer 2: should we link to it here?
```suggestion
But no more. Today, we're ready to announce the general availability of a new option in dbt Cloud: [**"Keep on latest version."**](/docs/dbt-versions/upgrade-dbt-version-in-cloud#keep-on-latest-version) 
```

Author: yes!

## How we safely deploy dbt upgrades to Cloud

We've put in place a rigorous, best-in-class suite of tests and control mechanisms to ensure that all changes to dbt under the hood are fully vetted before they're deployed to customers of dbt Cloud.

This pipeline has in fact been in place since January! It's how we've already been rolling out patch releases for existing dbt Core versions to dbt Cloud customers, and shipping continuous changes to the hundreds of customers who've selected "Keep on latest version" while it's been in Beta and Preview. In that time, this process has enabled us to prevent multiple regressions before they were rolled out to any customers.

We're very confident in the robustness of this process**. We also know that we'll need to continue building trust with time.** We're sharing details about this work in the spirit of transparency and to build that trust.

Any new change to dbt-core and adapters goes through the following steps before it's available to customers in dbt Cloud:

> <Lightbox src="/img/blog/2024-05-02-latest-dbt/testing-deploy-pipeline.png" title="Testing and deploy pipeline" />

Reviewer 2: ```suggestion
<Lightbox src="/img/blog/2024-05-22-latest-dbt/testing-deploy-pipeline.png" title="Testing and deploy pipeline" />
```

Reviewer 2: screenshot wasn't rendering right
<img width="665" alt="Screenshot 2024-05-02 at 14 59 32" src="https://github.com/dbt-labs/docs.getdbt.com/assets/89008547/a6d3c125-b65c-4e20-8df5-24d298f626dd">

Reviewer 2: committing this directly

> <Lightbox src="/img/blog/2024-05-22-latest-dbt/testing-deploy-pipeline.png" title="Testing and deploy pipeline" />

Reviewer 2: fixed this up bc it was rendering weird
<img width="665" alt="Screenshot 2024-05-02 at 14 59 32" src="https://github.com/dbt-labs/docs.getdbt.com/assets/89008547/8fa00bc1-be3a-4fdf-bd6f-a43ff32474e9">

> ### **Step 1: Unit & Functional Tests in dbt Core + Adapters**

Reviewer 2: ```suggestion
### **Step 1: Unit & functional tests in dbt Core + adapters**
```

First up is a battery of thousands of tests that we run dozens of times per day. No change, in either dbt-core or in the [data platform adapters](https://docs.getdbt.com/docs/trusted-adapters) supported by dbt Cloud, is merged until it has passed this full suite of tests. 

> Here, *unit tests* test internal components in isolation from one another, and *functional tests* represent edge cases in expected behaviour under known conditions.

Reviewer 2: ```suggestion
Here, *unit tests* test internal components in isolation from one another, and *functional tests* represent edge cases in expected behavior under known conditions.
```

For adapters, tests also ensure that the full matrix of data platform features continue to work as expected: BigQuery partitioning + incremental strategies, Snowflake data types + model contracts, Redshift sort keys — so on and so forth.

### Step 2: P**re-deploy smoke tests**

> Next, we create a Docker image with the latest dbt changes installed alongside each adapter supported in dbt Cloud. We run an additional suite of end-to-end tests on this image across the matrix of supported adapters, across a sample of projects and packages representing real-world complexity, through several dbt workflows. In doing so, this phase of testing also ensures that the latest version of dbt does not break compatibility with frequently relied-upon dbt packages. 

Reviewer: ```suggestion
Next, we create a Docker image with the latest dbt changes installed alongside each adapter supported in dbt Cloud. We run an additional suite of end-to-end tests on this image across a matrix of supported adapters, test projects that representing real-world complexity and typical dbt user workflows. In doing so, this phase of testing also ensures that the latest version of dbt does not break compatibility with frequently relied-upon dbt packages. 
```

Author: good suggestion - I'm going to keep "packages" in here because it's something we've been working on most recently, and because it comes up often in conversations

This breadth of testing provides early detection of any regressions that might have been introduced by our changes to dbt-core, changes by adapter maintainers, or any of their dependencies and drivers — using the exact installed versions that would be deployed to dbt Cloud. Crucially, this helps safeguard us from breaking changes in third-party software. 

### Step 3: **dbt Cloud service tests**

Before the new image version goes live, we ensure that all dbt changes are cross-compatible with every dbt Cloud service that depends on Core functionality, including areas such as the Cloud IDE, the Cloud CLI, scheduled job runs, CI, and connection testing. 

> For each dbt Cloud service we run a testing suite that consists of:

Reviewer 2: ```suggestion
For each dbt Cloud service, we run a testing suite that consists of:
```

- Unit and integration specific tests to behavior of each dbt Cloud service
- End-to-end headless browser testing for our UI-heavier applications
- Compatibility for each adapter with that service

> This step provides further depth in testing the interplay between dbt Core and dbt Cloud application-specific functionality, covering cases such as linting SQL that has an ephemeral reference, or resolving cross-project refs across multi-project "dbt Mesh" deployments.

Reviewer 2: ```suggestion
This step provides further depth in testing the interplay between dbt Core and dbt Cloud application-specific functionality, covering cases such as linting SQL that has an ephemeral reference, or resolving [cross-project refs](/docs/collaborate/govern/project-dependencies#how-to-write-cross-project-ref) across multi-project "dbt Mesh" deployments.
```

### Step 5: Phased **rollout**

Once the canary deployment has been proven to run stably for at least 24 hours, we mark it as eligible for all accounts to upgrade in their next scheduled deployment of dbt Cloud.

<aside>
ℹ️ Even with a robust testing, deployment, and monitoring system in place, it will never be *impossible* for a breaking change to make it through — just as in any other SaaS application.

If this does happen, we commit to identifying and rolling back any breaking changes as quickly as possible. Under the new testing and deployment model in dbt Cloud, we have consistently demonstrated to have been able roll back changes in less than an hour and ship roll forward bug fixes in less than 4 hours.

All incidents are retrospected to make sure we not only identify and fix the root cause(s), but also promptly put in place testing, automation, and quality gates to ensure that the same problem never happens again.

> </aside>

Author: @mirnawong1 Do you have an idea of how we could make this look nicer? Maybe an `:::info` callout? (I'll give this a try in a little bit)

The outcome of this process is that, when you select "Keep on latest version" in dbt Cloud, the time between an improvement or fix being made to dbt Core and you *safely* getting access to it in your projects is a matter of days — rather than months of waiting for the next dbt Core release, on top of any additional time it may have taken to actually carry out the upgrade.

We're pleased to say that since the beta launch of "Keep on latest version" in dbt Cloud, **we have not had any functional regressions reach customers**, while we've also been shipping 5-10 improvements to core dbt functionality every day. This is a foundation that we aim to build on for the foreseeable future.

## Stability as a feature

A rigorous testing pipeline in dbt Cloud is crucial, but real ongoing stability required some deeper changes in the dbt framework. We take our responsibility as the maintainers of dbt Core seriously, as well the open-source ecosystem around it.

> We've ******taken a longer release cycle for the upcoming 1.8 release of dbt Core to revisit some of the "do later" design choices we made in the past — specifically around adapter compatibility, behavior change management, and metadata artifacts.

Reviewer 2: not sure if we want asterisks here? or are we bolding?
```suggestion
We've taken a longer release cycle for the upcoming 1.8 release of dbt Core to revisit some of the "do later" design choices we made in the past — specifically around adapter compatibility, behavior change management, and metadata artifacts.
```

Author: good catch - this is just a weird vestige of notion -> markdown export

### **Decoupling the adapter interface**

The adapter interface — i.e. how dbt Core actually connects to a third-party data platform — has historically been somewhat of a pain point. Adapter maintainers have often been *required to make* reactive changes when there's been an update to dbt Core.

> To solve that, we've released a new set of interfaces that are entirely independent of the `dbt-core` library: `[dbt-adapters==1.0.0](https://github.com/dbt-labs/dbt-adapters)`. From now on, any changes to `dbt-adapters` will be backward and forward-compatible. This also decouples adapter maintenance from the regular release cadence of `dbt-core`  — meaning maintainers get full control over when they ship implementations of new adapter-powered features.

Reviewer 2: the link isn't rendering right so fixing this
```suggestion
To solve that, we've released a new set of interfaces that are entirely independent of the `dbt-core` library: [`dbt-adapters==1.0.0`](https://github.com/dbt-labs/dbt-adapters). From now on, any changes to `dbt-adapters` will be backward and forward-compatible. This also decouples adapter maintenance from the regular release cadence of `dbt-core`  — meaning maintainers get full control over when they ship implementations of new adapter-powered features.
```

Note that adapters running in dbt Cloud **must** be [migrated to the new decoupled architecture](https://github.com/dbt-labs/dbt-adapters/discussions/87) as a baseline in order to support the new "Keep on latest version".

> ### **Managing behaviour changes: stability as a feature**

Reviewer 2: ```suggestion
### **Managing behavior changes: stability as a feature**
```

> We all want the benefits of a stable, actively maintained product. Occasionally the dbt Labs team identifies a change to default behavior that we believe is more sensible, more secure, more helpful — just better in some way — but which would come as a change to users who have grown accustomed to the existing behaviour.

Reviewer 2: ```suggestion
We all want the benefits of a stable, actively maintained product. Occasionally the dbt Labs team identifies a change to default behavior that we believe is more sensible, more secure, more helpful — just better in some way — but which would come as a change to users who have grown accustomed to the existing behavior.
```

> To accommodate both groups in these scenarios, we've extended dbt to support project-level behaviour  flags. These can be used to *opt into* or *opt out of* changes to default behaviour. From now on, backward-incompatible changes to dbt functionality will be implemented behind a flag with a default value that preserves the legacy behaviour. After a few months, the new behaviour will become the default — but only after some proactive communication with customers and external package maintainers.

Reviewer 2: ```suggestion
To accommodate both groups in these scenarios, we've extended dbt to support project-level behavior flags. These can be used to *opt into* or *opt out of* changes to default behavior. From now on, backward-incompatible changes to dbt functionality will be implemented behind a flag with a default value that preserves the legacy behavior. After a few months, the new behavior will become the default — but only after some proactive communication with customers and external package maintainers.
```

> The same behaviour change flags will naturally extend to dbt packages, which are fundamentally just dbt projects. This allows package maintainers to ensure that behaviour doesn't change unexpectedly as a result of changes to dbt Core. For more details, check out our user documentation on [legacy behaviours](https://docs.getdbt.com/reference/global-configs/legacy-behaviors#behaviors), as well as our [contributor documentation](https://github.com/dbt-labs/dbt-core/blob/main/docs/eli64/behavior-change-flags.md) for introducing behaviour changes safely.

Reviewer 2: ```suggestion
The same behavior change flags will naturally extend to dbt packages, which are fundamentally just dbt projects. This allows package maintainers to ensure that behavior doesn't change unexpectedly as a result of changes to dbt Core. For more details, check out our user documentation on [legacy behaviors](https://docs.getdbt.com/reference/global-configs/legacy-behaviors#behaviors), as well as our [contributor documentation](https://github.com/dbt-labs/dbt-core/blob/main/docs/eli64/behavior-change-flags.md) for introducing behavior changes safely.
```

**Stability for metadata artifacts**

Lastly, we've revisited our [artifact interfaces](https://docs.getdbt.com/reference/artifacts/dbt-artifacts) — which are the workhorses of many integrations in the dbt ecosystem, whether maintained by dbt Labs, third-party tools and vendors, or are custom to a particular organization. While these schemas have adhered to a well-defined and versioned since dbt-core 1.0, they have historically seen major changes on almost every minor release of dbt-core. We've [formalized our development best practices](https://github.com/dbt-labs/dbt-core/blob/main/core/dbt/artifacts/README.md#making-changes-to-dbtartifacts) on strongly preferring minor schema evolutions over major breaking changes, and put [automated CI in place](https://github.com/dbt-labs/dbt-core/blob/main/.github/workflows/check-artifact-changes.yml) to ensure we are not unintentionally introducing breaking changes to artifacts to avoid disruption to artifact-based integrations across the OSS ecosystem.

---

> In conclusion, we're putting new muscle behind our commitments to the community & ecosystem:

Reviewer 2: sugg a h2 to finish it out
```suggestion
## Conclusion

In conclusion, we're putting new muscle behind our commitments to the community and ecosystem:
```

- Compatibility for adapters, maintained by us and by others

> - Predictable behaviour changes for maintainers of project and package code

Reviewer 2: ```suggestion
- Predictable behavior changes for maintainers of project and package code
```

## Our commitment

In conclusion, we’re putting a lot of new muscle behind our commitments to dbt Cloud customers, the dbt Community, and the broader ecosystem:

- **Continuous updates**: "Keep on latest version" in dbt Cloud simplifies the update process, ensuring you always have the latest features and bug fixes without the maintenance overhead.
- **A rigorous new testing and deployment process**: Our new testing pipeline ensures that every update is carefully vetted against documented interfaces, Cloud-supported adapters, and popular packages before it reaches you. This process minimizes the risk of regressions — and has now been successful at entirely preventing them for hundreds of customers over multiple months.
- **A commitment to stability**: We’ve reworked our approaches to adapter interfaces, behaviour change management, and metadata artifacts to give you more stability and control.

> As we continue to enhance dbt Cloud, our commitment remains firm: to provide a stable, dependable platform that allows our users to spend less time on maintenance overhead and focus more on disseminating organizational knowledge that ~~drives business value for their embedded stakeholders who are definitely not going to be replaced by LLMs~~ creating value.

Reviewer 3: ```suggestion
As we continue to enhance dbt Cloud, our commitment remains firm: to provide a stable, dependable platform that allows our users to spend less time on maintenance overhead and focus more on disseminating organizational knowledge that ~~drives business value for their embedded stakeholders who are definitely not going to be replaced by LLMs~~ creates value.
```
